pkg/webhook/preflight/nutanix/failuredomain.go

```diff
@@ -0,0 +1,220 @@
+		Allowed: true,
+	}
+
+	// Fetch the referent failure domain object
+	fdObj := &capxv1.NutanixFailureDomain{}
+	fdKey := ctrlclient.ObjectKey{Name: fdc.failureDomainName, Namespace: fdc.namespace}
+	if err := fdc.kclient.Get(ctx, fdKey, fdObj); err != nil {
+		if errors.IsNotFound(err) {
+			result.Allowed = false
+			result.Causes = append(result.Causes, preflight.Cause{
+				Message: fmt.Sprintf(
+					"NutanixFailureDomain %q referenced in cluster was not found in the management cluster. Please create it and retry.", //nolint:lll // Message is long.
+					fdc.failureDomainName,
+				),
+				Field: fdc.field,
+			})
+			return result
+		}
+
+		result.Allowed = false
+		result.InternalError = true
+		result.Causes = append(result.Causes, preflight.Cause{
+			Message: fmt.Sprintf(
+				"Failed to get NutanixFailureDomain %q: %v This is usually a temporary error. Please retry.", //nolint:lll // Message is long.
+				fdc.failureDomainName,
+				err,
+			),
+			Field: fdc.field,
+		})
+		return result
+	}
+
+	// Validate the failure domain configuration
+	// Validate spec.prismElementCluster configuration
+	peIdentifier := fdObj.Spec.PrismElementCluster
+	peClusters, err := getClusters(fdc.nclient, &peIdentifier)
+	if err != nil {
+		result.Allowed = false
+		result.InternalError = true
+		result.Causes = append(result.Causes, preflight.Cause{
+			Message: fmt.Sprintf(
+				"Failed to check if the Prism Element cluster %q, referenced by Failure Domain %q, exists: %v This is usually a temporary error. Please retry.", //nolint:lll // Message is long.
+				peIdentifier,
+				fdc.failureDomainName,
+				err,
+			),
+			Field: fdc.field,
+		})
+		return result
+	}
+	if len(peClusters) != 1 {
+		result.Allowed = false
+		result.Causes = append(result.Causes, preflight.Cause{
+			Message: fmt.Sprintf(
+				"Found %d Prism Element cluster(s) that match identifier %q. There must be exactly 1 that matches this identifier.", //nolint:lll // Message is long.
+				len(peClusters),
+				peIdentifier,
+			),
+			Field: fdc.field,
+		})
+		return result
+	}
+	peUUID := *peClusters[0].ExtId
+
+	// Validate spec.subnets configuration
+	for _, id := range fdObj.Spec.Subnets {
+		subnets, err := getSubnets(fdc.nclient, &id)
+		if err != nil {
+			result.Allowed = false
+			result.InternalError = true
+			result.Causes = append(result.Causes, preflight.Cause{
+				Message: fmt.Sprintf(
+					"Failed to get subnet %q referenced by the Failure Domain %q: %v This is usually a temporary error. Please retry.", //nolint:lll // Message is long.
+					id,
+					fdc.failureDomainName,
+					err,
+				),
+				Field: fdc.field,
+			})
+			continue
+		}
+
+		// Filter the subnets, the valid ones should either have clusterReference match peUUID
+		// or clusterReference being nil (for overlay subnets)
+		filteredSubnets := []netv4.Subnet{}
+		for i := range subnets {
+			if subnets[i].ClusterReference == nil || *subnets[i].ClusterReference == peUUID {
```

Comment:
if multiple subnets have the same name and are overlay subnets thats not a valid configuration?

Reply:
Currently we only check the subnet with the specified name exists and only one in PC, this includes the subnets with the clusterReference and those overlays. If not found or more than one found, this check will report error. We don't check if the subnet configuration in PC is correctly.